New durability section - in development (#4677)
EnterpriseDB/docs: product_docs/docs/pgd/5/durability/limitations.mdx

    - `CLUSTER` (of a single relation or index only)
    - `ANALYZE`
    - `TRUNCATE`

- Explicit two-phase commit is not supported by Group Commit as it already uses two-phase commit.

- Combining different commit decision options in the same transaction or
  combining different conflict resolution options in the same transaction isn't
  supported.

- Currently, Raft commit decisions are extremely slow, producing very low TPS.
  We recommended using them only with the `eager` conflict resolution setting
  to get the Eager All-Node Replication behavior of PGD 4 and older.

## Eager

[Eager](../consistency/eager) is available through Group Commit. It avoids conflicts by eagerly aborting transactions that might clash. It's subject to the same limitations as Group Commit.

- Eager doesn't allow the `NOTIFY` SQL command or the `pg_notify()` function. It
  also don't allow `LISTEN` or `UNLISTEN`.

## CAMO

[Commit At Most Once](camo) (CAMO) is a feature that aims to prevent
applications committing more than once. If you use this feature, take 
these limitations into account when planning:

- CAMO is designed to query the results of a recently failed COMMIT on the

> origin node. So, in case of disconnection, code the application to immediately
> request the transaction status from the CAMO partner. Have as little delay as
> possible after the failure before requesting the status. Applications must not
> rely on CAMO decisions being stored for longer than 15 minutes.

Reviewer: ```suggestion
origin node. In case of a connection failure, the application must
request the transaction status from the CAMO partner immediately. Ensure to have as little delay as
possible after the failure before requesting the status. Applications must not
rely on CAMO decisions being stored for longer than 15 minutes.
```

---
title: Limitations
---

The following limitations apply to the use of commit scopes and the various durability options they enable.

Group Commit transactions can't yet execute DDL, and they don't support [explicit
two-phase commit](../twophase). Future releases might allow them. However, the `TRUNCATE`
command is allowed.

You can combine only CAMO transactions with the `DEGRADE TO` clause for
switching to asynchronous operation in case of lowered availability.

Group Commit, Eager and CAMO transactions aren't currently supported in
combination with the Decoding Worker feature or with transaction streaming.
Installations using these commit scopes must keep `enable_wal_decoder` and `streaming_mode`
disabled for the PGD node group.

[Legacy synchronous replication](legacy-sync) uses a mechanism for transaction
confirmation different from Group Commit. The two aren't compatible, and you
must not use them together. Therefore, whenever you Group Commit transactions,

> make sure none of the PGD nodes are configured in `synchronous_standby_names`.

Reviewer: Probably a question for @PJMODOS or @mwanner - could we technically prevent this by raising an error if the user enables legacy synchronous replication?
